lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -775,4 +795,156 @@ toLinearLayout(ArrayRef<int64_t> shape, Attribute layout,
+  //             let shmem[offset] in block b contain output element `out`.
+  //             offset += 1
+  //             seen.add(out)
+  //
+  // Doing this is simpler than it may seem.  There are some bits in the inputs
+  // that don't affect the output; these are the "free variables".  Our layout
+  // simply sets these to 0, and then everything works as we want.
+  //
+  // TODO(jlebar): Currently we ignore dst entirely; we're just choosing
+  // something which is good for src.
+  llvm::MapVector<StringAttr, int> freeVars = src.getFreeVariableMasks();
+  LinearLayout::BasesT bases;
+
+  // Ensure kOffset is present even if the whole src layout is free variables
+  // (meaning there's only one unique element per block).
+  bases.insert({kOffset, {}});
+
+  for (StringAttr inDim : src.getInDimNames()) {
+    if (inDim == kBlock) {
+      continue;
+    }
+    assert(freeVars.contains(inDim));
+    int32_t dimFreeVars = freeVars[inDim];
+    for (int i = 0; i < src.getInDimSizeLog2(inDim); i++) {
+      bool isFree = dimFreeVars & (1 << i);
+      if (isFree) {
+        continue;
+      }
+
+      llvm::MapVector<StringAttr, int> inIdx;
+      for (StringAttr d : src.getInDimNames()) {
+        inIdx[d] = 0;
+      }
+      inIdx[inDim] = 1 << i;
+
+      auto outIdx = to_vector(make_second_range(src.apply(inIdx.takeVector())));
+      bases[kOffset].push_back(
+          std::vector<int32_t>(outIdx.begin(), outIdx.end()));
+    }
+  }
+
+  // The `block` bases match the input layout exactly.  This is so there's no
+  // cross-CTA transfers.
+  bases[kBlock] = src.getBases().lookup(kBlock);
+  LinearLayout singleIter(std::move(bases), to_vector(src.getOutDimNames()));
+
+  // Now split up the layout into one or more iterations.  We split it up so we
+  // do one vectorized access per iteration.
+  LinearLayout storeLayout = src.invertAndCompose(singleIter);
+  int vec = std::min(maxVecElems, storeLayout.getNumConsecutiveInOut());
+  int numIters = storeLayout.getInDimSize(kRegister) / vec;
+
+  // Put the first `vec` elements into the first iteration, and so on.
+  LinearLayout ret =
+      singleIter
+          .reshapeIns({
+              {kOffset, singleIter.getInDimSize(kOffset) / numIters},
+              {kIteration, numIters},
+              {kBlock, singleIter.getInDimSize(kBlock)},
+          })
+          .transposeIns({kOffset, kBlock, kIteration});
+
+  return ret;
+}
+
+LinearLayout
+chooseShemLayoutForRegToRegConversion(MLIRContext *ctx,
+                                      ArrayRef<unsigned> tensorShape,
+                                      ArrayRef<unsigned> repShape) {
+  auto outDimNames = standardOutDimNames(ctx, tensorShape.size());
+  LinearLayout layout = LinearLayout::empty();
+  SmallVector<StringAttr> kRepDims;
+  SmallVector<StringAttr> kOffsetDims;
+  auto totalIters = 1;
+  auto totalOffsets = 1;
+  for (int i = 0; i < tensorShape.size(); i++) {
+    // Start with the fastest changing dimension.
```

Comment:
how do we know this is the fastest-changing dimension?  Doesn't that depend on the *layout* of the shape?

Reply:
It refers to the intermedidate tensor. The `src` tensor is accessed based on the layout of `src`, but the intermediate tensor is accessed based on the layout we created here.
For example in the 2d case (dim0, dim1), we always access following the row-major order, which is dim1 -> dim0.